rfd/0003-deduplication.md

```diff
@@ -0,0 +1,214 @@
+should probably not provide this to the boefje. It should be easy to refactor
+this so it used the serialized input OOI.
+
+#### Usage of environment field
+
+This is not used directly from the BoefjeMeta, but this
+
+#### Usage of started_at, endated_at, boefje and runnable_hash fields
+
+This isn't used by any boefje and can be safely removed.
+
+#### Boefje signature / arguments
+
+Boefje that does not require organization can have the following signature:
+
+```python
+def run(input_ooi: dict[str, Any]) -> list[tuple[set, bytes | str]]:
+```
+
+And boefje that requires organization:
+
+```python
+def run(input_ooi: dict[str, Any], organization: str) -> list[tuple[set, bytes | str]]:
+```
+
+The `input_ooi` is the serialized input OOI that is stored in the BoefjeMeta
+`arguments`. The environment field does need to be passed as argument because it
+is already passed to the boefje using environment variables.
+
+### Deduplication implementation
+
+The scheduler is the place where boefjes are scheduled so that seems to be most
+obvious place to implement deduplication. Because the data necessary is
+currently spread out over multiple services / databases this means it would need
+to do requests to other services to fetch this information. This is unavoidable
+because there is currently no single place with all the information.
+
+#### Whether boefje needs organization
+
+The scheduler already fetches and caches all the available boefjes for an
+organization, so the scheduler can use this data to easily check if the boefje
+requires the organization and can't be deduplicated.
+
+#### Environment settings from previous task in other organization
+
+This data is stored in bytes and needs to be fetched by the scheduler to
+determine whether the environment settings are the same.
+
+#### Environment settings of to be deduplicated boefje task
+
+This is a complicated. The scheduler can fetch the current configured
+environment from the katalogus, but this is merged with the environment that is
+passed directly to the boefje runner as environment variables. There is
+currently no easy way to fetch those and adding an API for this to the boefje
+runner doesn't seem to be a good idea.
+
+The best course of action seems to be to disallow the environment variables
+passed to the boefje runner when deduplication is turned on. If both
+deduplication is turned on and environment variables are passed to the boefje
+runner we need to return an error because the user needs to know that the passed
+variables aren't used.
+
+#### Deduplication algorithm
+
+A possible algorithm for deduplication could be:
+
+- The scheduler wants to scheduler a boefje for a specific organization and
+  input OOI.
+- The scheduler checks if the boefje needs the organization. If so it can
+  continu with scheduling the boefje task.
+- The scheduler queries the stored tasks to see if a different organization has
+  already run on this input OOI.
```

Comment:
What is the time window for fetching those tasks? Since no job will rerun at all strictly reading this algorithm

Reply:
Max age of possible raw files should be the interval on the requesting organisations boefje. So if the requestion org has an interval of 1 day, files no older than 1 day could be used from another organisation.